numpy/core/src/umath/loops_unary_fp.dispatch.c.src

```diff
@@ -0,0 +1,276 @@
+/*@targets
+ ** $maxopt baseline
+ ** sse2 vsx2 neon
+ **/
+/**
+ * Force use SSE only on x86, even if AVX2 or AVX512F are enabled
+ * through the baseline, since scatter(AVX512F) and gather very costly
+ * to handle non-contiguous memory access comparing with SSE for
+ * such small operations that this file covers.
+*/
+#define NPY_SIMD_FORCE_128
+#define NPY_NO_EXPORT NPY_VISIBILITY_HIDDEN
+#include "numpy/npy_math.h"
+#include "simd/simd.h"
+#include "loops.h"
+/*
+ * nomemoverlap - returns true if two strided arrays have an overlapping
+ * region in memory. ip_size/op_size = size of the arrays which can be negative
+ * indicating negative steps.
+ */
+NPY_FINLINE int
+nomemoverlap(char *ip, npy_intp ip_size, char *op, npy_intp op_size)
+{
+    char *ip_start, *ip_end, *op_start, *op_end;
+    if (ip_size < 0) {
+        ip_start = ip + ip_size;
+        ip_end = ip;
+    }
+    else {
+        ip_start = ip;
+        ip_end = ip + ip_size;
+    }
+    if (op_size < 0) {
+        op_start = op + op_size;
+        op_end = op;
+    }
+    else {
+        op_start = op;
+        op_end = op + op_size;
+    }
+    return (ip_start > op_end) | (op_start > ip_end);
+}
```

Comment:
```suggestion
NPY_FINLINE int
loops_is_overlap(const void *src, npy_intp src_step, const void *dst, npy_intp dst_step, int unroll_len)
{
    if (src_step && dst_step) {
        char *src_step1 = (char*)src + src_step * unroll_len;
        char *dst_step1 = (char*)dst + dst_step * unroll_len;
        if (dst_step1 > src_step1) {
            return (dst_step1 - src_step1) < (src_step * unroll_len);
        }
        if (dst_step1 < src_step1) {
            return (src_step1 - dst_step1) < (dst_step * unroll_len);
        }
    }
    return 0;
}
```
This function is widely used in [simd.inc](https://github.com/numpy/numpy/blob/master/numpy/core/src/umath/simd.inc.src), and it has a negative effect on the performance.

I don't know why we check the whole array size rather than the unrolled size?
also, this function returns false if the two arrays mapping to each other(same line).

Reply:
/cc @r-devulap

Reply:
I went solo here and implmented `loops_is_overlap()` to be used instead of `nomemoverlap()`

Reply:
@seiko2plus This was added in this PR: https://github.com/numpy/numpy/pull/15612. `np.maximum.accumulate` results in memory overlap for input and output arrays. The idea was to avoid using SIMD if there is any overlap between input and output arrays (accounting for -ve strides too).

Reply:
@r-devulap,
> The idea was to avoid using SIMD if there is any overlap between input and output arrays (accounting for -ve strides too).

I'm aware of the idea behind `nomemoverlap()`, my concern is to reduce the reduction by
 - allow aliasing (same pointer & step size)
 - only check the unroll size(SIMD width) instead of the whole array size

Reply:
> allow aliasing (same pointer & step size)

should be fine AFAIK. 

> only check the unroll size(SIMD width) instead of the whole array size

It might be sufficient, but I am not the best person to say that for sure. I went with the most paranoid assumptions because I wanted to be damn sure to cover all possible scenarios. For example, base pointers of` in `and `out `arrays are far off, but `in` array has a positive stride and `out `array has a negative stride and they overlap towards the tail. This scenario may not ever happen but I wasn't 100% certain.

Reply:
marked as resolved according to https://github.com/numpy/numpy/pull/16247#discussion_r511497632